File: 2024/day18/zig/solution.zig
const std = @import("std");
const ArrayList = std.ArrayList;

const SIZE: usize = 71;
const NUM_BYTES: usize = 1024;

const Point = struct {
    x: i32,
    y: i32,

    fn toIndex(self: Point) usize {
        return @as(usize, @intCast(self.y)) * SIZE + @as(usize, @intCast(self.x));
    }

    fn isValid(self: Point) bool {
        return self.x >= 0 and self.x < SIZE and self.y >= 0 and self.y < SIZE;
    }
};

const QueueItem = struct {
    point: Point,
    steps: i32,
};

const directions = [_]Point{
    .{ .x = 0, .y = 1 },
    .{ .x = 0, .y = -1 },
    .{ .x = 1, .y = 0 },
    .{ .x = -1, .y = 0 },
};

fn bfs(corrupted: []const bool, allocator: std.mem.Allocator) !?i32 {
    const start = Point{ .x = 0, .y = 0 };
    const goal = Point{ .x = SIZE - 1, .y = SIZE - 1 };

    if (corrupted[start.toIndex()] or corrupted[goal.toIndex()]) {
        return null;
    }

    var visited: [SIZE * SIZE]bool = [_]bool{false} ** (SIZE * SIZE);
    visited[start.toIndex()] = true;

    var queue: ArrayList(QueueItem) = .empty;
    defer queue.deinit(allocator);

    try queue.append(allocator, .{ .point = start, .steps = 0 });

    var read_idx: usize = 0;
    while (read_idx < queue.items.len) {
        const item = queue.items[read_idx];
        read_idx += 1;

        if (item.point.x == goal.x and item.point.y == goal.y) {
            return item.steps;
        }

        for (directions) |dir| {
            const nx = item.point.x + dir.x;
            const ny = item.point.y + dir.y;
            const next = Point{ .x = nx, .y = ny };

            if (next.isValid()) {
                const idx = next.toIndex();
                if (!visited[idx] and !corrupted[idx]) {
                    visited[idx] = true;
                    try queue.append(allocator, .{ .point = next, .steps = item.steps + 1 });
                }
            }
        }
    }

    return null;
}

fn part1(positions: []const Point, allocator: std.mem.Allocator) !?i32 {
    var corrupted: [SIZE * SIZE]bool = [_]bool{false} ** (SIZE * SIZE);

    const limit = @min(NUM_BYTES, positions.len);
    for (positions[0..limit]) |pos| {
        if (pos.isValid()) {
            corrupted[pos.toIndex()] = true;
        }
    }

    return try bfs(&corrupted, allocator);
}

fn part2(positions: []const Point, allocator: std.mem.Allocator) !Point {
    var left: usize = 0;
    var right: usize = positions.len;

    while (left < right) {
        const mid = (left + right) / 2;

        var corrupted: [SIZE * SIZE]bool = [_]bool{false} ** (SIZE * SIZE);
        for (positions[0 .. mid + 1]) |pos| {
            if (pos.isValid()) {
                corrupted[pos.toIndex()] = true;
            }
        }

        const result = try bfs(&corrupted, allocator);
        if (result == null) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }

    return positions[left];
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Parse positions
    var positions: ArrayList(Point) = .empty;
    defer positions.deinit(allocator);

    var lines = std.mem.splitScalar(u8, content, '\n');
    while (lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, &[_]u8{ ' ', '\r', '\t' });
        if (trimmed.len == 0) continue;

        var parts = std.mem.splitScalar(u8, trimmed, ',');
        const x_str = parts.next() orelse continue;
        const y_str = parts.next() orelse continue;

        const x = std.fmt.parseInt(i32, x_str, 10) catch continue;
        const y = std.fmt.parseInt(i32, y_str, 10) catch continue;

        try positions.append(allocator, .{ .x = x, .y = y });
    }

    // Part 1
    const p1_result = try part1(positions.items, allocator);
    std.debug.print("Part 1: {d}\n", .{p1_result.?});

    // Part 2
    const p2_result = try part2(positions.items, allocator);
    std.debug.print("Part 2: {d},{d}\n", .{ p2_result.x, p2_result.y });
}

File: 2024/day18/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	gridSize     = 71
	initialBytes = 1024
)

type Point struct {
	x, y int
}

type QueueItem struct {
	point Point
	steps int
}

func parseInput(filename string) ([]Point, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var positions []Point
	scanner := bufio.NewScanner(file)
	lineNum := 0
	for scanner.Scan() {
		lineNum++
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) != 2 {
			return nil, fmt.Errorf("line %d: expected format 'x,y', got %q", lineNum, line)
		}
		x, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("line %d: parsing x coordinate: %w", lineNum, err)
		}
		y, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("line %d: parsing y coordinate: %w", lineNum, err)
		}
		positions = append(positions, Point{x, y})
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("reading input: %w", err)
	}
	return positions, nil
}

// bfs finds the shortest path from (0,0) to (size-1,size-1) avoiding corrupted cells.
// Returns (steps, true) if a path exists, or (0, false) if no path is found.
func bfs(corrupted map[Point]bool, size int) (int, bool) {
	start := Point{0, 0}
	goal := Point{size - 1, size - 1}

	if corrupted[start] || corrupted[goal] {
		return 0, false
	}

	queue := []QueueItem{{start, 0}}
	visited := make(map[Point]bool)
	visited[start] = true

	directions := []Point{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.point == goal {
			return current.steps, true
		}

		for _, d := range directions {
			next := Point{current.point.x + d.x, current.point.y + d.y}
			if next.x >= 0 && next.x < size && next.y >= 0 && next.y < size &&
				!visited[next] && !corrupted[next] {
				visited[next] = true
				queue = append(queue, QueueItem{next, current.steps + 1})
			}
		}
	}

	return 0, false
}

func part1(positions []Point, numBytes, size int) int {
	corrupted := make(map[Point]bool)
	for i := 0; i < numBytes && i < len(positions); i++ {
		corrupted[positions[i]] = true
	}
	steps, _ := bfs(corrupted, size)
	return steps
}

func part2(positions []Point, size int) string {
	left, right := 0, len(positions)

	for left < right {
		mid := (left + right) / 2
		corrupted := make(map[Point]bool)
		for i := 0; i <= mid; i++ {
			corrupted[positions[i]] = true
		}
		_, found := bfs(corrupted, size)
		if !found {
			right = mid
		} else {
			left = mid + 1
		}
	}

	blockingPos := positions[left]
	return fmt.Sprintf("%d,%d", blockingPos.x, blockingPos.y)
}

func main() {
	positions, err := parseInput("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Part 1:", part1(positions, initialBytes, gridSize))
	fmt.Println("Part 2:", part2(positions, gridSize))
}
